### ISSUE TITLE
Turn the hint locality warning into an error

### ISSUE THREAD

**Issue by ppedrot** [coq/coq#13394]
After #13384 that introduced a warning for hint commands without an explicit locality in 8.13, this warning needs to be turned into an error by default in 8.14.

Next step is to remove the warning / error in 8.15 with the default locality set to `export`.

**Comment by Zimmi48** [comment]
As noticed by @silene following the CoqPL dev session, PR #13725 by @SkySkimmer extended the deprecation warning of hints without a locality to the `Hint Rewrite` command. This makes it impossible to apply the plan as is: we cannot turn this warning into an error now, otherwise users would get a bad experience for `Hint Rewrite` (which didn't even support locality attributes so far). If we want to keep the plan as is (instead of delaying it one or a few versions), we need to distinguish between the `Hint Rewrite` and the other `Hint` warnings.

**Comment by gares** [comment]
We can amend #13725 by using a different warning if this smooths the plan for the other

**Comment by ppedrot** [comment]
We're not in a hurry, though. If desired we can keep the same warning and postpone the error-by-default change. I believe (naively maybe?) that a warning is enough to deter most users from writing a hint without locality.

**Comment by Zimmi48** [comment]
I prefer this plan: it would be cleaner to give more time to users and switch the whole Hint mechanism at once.

**Comment by Zimmi48** [comment]
> Next step is to remove the warning / error in 8.15 with the default locality set to `export`.

Note that in #13562, it was proposed to change the default not to `export` but to `local` so that the default locality would be the same inside and outside sections.

**Comment by Zimmi48** [comment]
> We're not in a hurry, though. If desired we can keep the same warning and postpone the error-by-default change. I believe (naively maybe?) that a warning is enough to deter most users from writing a hint without locality.

I'm tentatively moving this to the 8.15 milestone to reflect this change of plan (this change of milestone can be reverted of course, but it would require amending #13725 to use a different warning).

**Comment by RalfJung** [comment]
Is `Hint Rewrite` truly the only `Hint` command that was missing, or are there still others that would need to gain support for locality attributes in the first place?

**Comment by Zimmi48** [comment]
If you do not include `Instance` in the list of `Hint` commands (as it doesn't support `export` yet), then yes that was the last missing `Hint` command.

**Comment by RalfJung** [comment]
Yeah I mean commands whose syntax is `Hint .*`.

**Comment by RalfJung** [comment]
E.g. I just noticed there's `Hint View`; it's not clear [from the docs](https://coq.inria.fr/refman/proof-engine/ssreflect-proof-language.html#coq:cmd.Hint-View-for) that locality modifiers are supported.

**Comment by Zimmi48** [comment]
Wow! That one is new to me. Indeed, it doesn't seem to support any locality attribute (and doesn't trigger any warning). @coq/ssreflect-maintainers: would you care to fix that?

**Comment by gares** [comment]
What exactly? The hint-view object is a global (not a superglobal) one so its default is OK if I'm not mistaken.

**Comment by RalfJung** [comment]
At least it should be possible to specify `Local` vs `Global` for `Hint View`.

**Comment by gares** [comment]
I think you should open an issue with this feature wish, if you have a use case, someone may implement it.

But just to be sure I understood things right: this does not impact the warning in question, we don't need to change the default locality of Hint View. Am I right? MC users won't get a warning/error for Hint View not being "localized", right?

**Comment by Zimmi48** [comment]
It is related to this issue because if we decide in the end that we want to make `local` the default locality (and not `export`), then it should change for this command as well (for consistency). See the [comment above](https://github.com/coq/coq/issues/13394#issuecomment-765472881):

> > Next step is to remove the warning / error in 8.15 with the default locality set to `export`.
> 
> Note that in #13562, it was proposed to change the default not to `export` but to `local` so that the default locality would be the same inside and outside sections.

So this mean that we should take a decision about this question sooner rather than later.

**Comment by Zimmi48** [comment]
> At least it should be possible to specify `Local` vs `Global` for `Hint View`.

I would avoid having `global` for one `Hint` command mean the same as `export` for another one. So I wouldn't add support for `global` to this command, only `export` and `local`.

**Comment by gares** [comment]
TTBOMK these two
- https://github.com/coq/coq/blob/master/library/libobject.mli#L153
- https://github.com/coq/coq/blob/master/plugins/ssr/ssrview.ml#L51

imply that `Hint View` is already export by default, and it should stay export by default IMO, these hints are always to be visible by your users, otherwise your views don't apply. Eg, I'd need to add (and keep) `#[export]` to all occurrences of the command, which is not really an improvement.

It also does not survive section closing, so it's "local" inside sections. I can add support for `#[local]` in modules if Ralf asks for it.

So, what should I do?

**Comment by RalfJung** [comment]
I am not using `Hint View`. I just think that all `Hint` should behave the same -- so I agree with @Zimmi48's comments above.

> Eg, I'd need to add (and keep) #[export] to all occurrences of the command, which is not really an improvement.

Well, maybe some of those were in sections and people forgot to `#[expot]` them, in that case it would be good to add the `#[export]`.

I think it is very confusing that the default is different inside a `Section` vs outside. Given that `Global`/`#[export]` is simply not supported for `Hint` inside `Section`, there are two ways to fix that:
* Make `Local` the default.
* Error when using an unqualified `Hint` in a section, telling the user to explicitly add `Local` to acknowledge that this `Hint` will not survive `End section`

I think either option would be much better than the current situation. It seems @gares would prefer the latter over the former.

**Comment by Zimmi48** [comment]
There's a third option (which IMHO is better than the second one) which is to add a warning to simply remind the user that the hint will not be exported.

**Comment by RalfJung** [comment]
Warnings are very easily overlooked (at least in PG), and this still means that the default is different inside vs outside sections.

**Comment by ppedrot** [comment]
I don't know why everybody is freaking out about `Hint View`. Despite the name, it shouldn't be considered a hint command. Problem solved.

**Comment by RalfJung** [comment]
@ppedrot from a user perspective, that's horrible. It's a `Hint` command, so as a user I'd expect it to work like all the other `Hint` commands. I shouldn't have to know about such implementation details to be able to predict `Hint` behavior.

**Comment by Zimmi48** [comment]
> Warnings are very easily overlooked (at least in PG), and this still means that the default is different inside vs outside sections.

It's up to the user to transform the warnings they care about into hard errors. Imposing `Local` inside sections if the only reason is that the user might have forgotten this was the default is not nice.

> @ppedrot from a user perspective, that's horrible. It's a `Hint` command, so as a user I'd expect it to work like all the other `Hint` commands. I shouldn't have to know about such implementation details to be able to predict `Hint` behavior.

The same thing was argued for `Hint Rewrite` before. It's too easy to say we don't care that we have inconsistent naming in our language.

**Comment by gares** [comment]
> I don't know why everybody is freaking out about `Hint View`. Despite the name, it shouldn't be considered a hint command. Problem solved.

> from a user perspective, that's horrible. It's a `Hint` command, so as a user I'd expect it to work like all the other `Hint` commands.

I understand both points of view, there is a real confusion here. Let me try to clarify.

The "hint locality" plan was about is not "the commands which have a `Hint` in their name" but rather "a subset of them which had the wrong, superglobal, default" (and maybe there are other superglobals out there, but are not part of this plan).

There is another "plan" here that is to make all commands which have a `Hint` in their name obey the same defaults.

Both are reasonable, but we should not mix them.

E.g. `Hint Views` have a reason to be named `Hint`, since they provide a directive for automation. Then name was not chosen considering a rule for locality, since there was no uniformity whatsoever back then. In the second incarnation of the code handling them, I called them "view adaptors". I would not object renaming the command to make things more uniform, there are very few occurrences out there.

Going back to this issue, which I believe talk about the superglobal plan. I think I should do nothing for `Hint View`. Do we agree on that?

**Comment by gares** [comment]
> It seems @gares would prefer the latter over the former.

For `Hint Views` yes.

I'm not an expert of the other hints. We use many `Hint Resolve` in MC, and they are all "export" I guess, unless when they are in a section and in that case are probably local. So I don't have any opinion on the right default for `Hint *`, assuming there is one.

**Comment by Zimmi48** [comment]
The reason to mix things is to try to see the big picture as much as possible before acting on any change.

In particular, the following objectives would be inconsistent:
- turn standard `Hint` commands local by default;
- keep `Hint View` command export by default;
- keep `Hint View` command name;
- uniformize commands that "look the same".

So we should decide to abandon one of these objectives. E.g., it could be the first one (which was never really debated up to now, so may be one of the easiest to abandon) or it could be the third one (Enrico seems to think this is doable).

**Comment by ppedrot** [comment]
>  It's too easy to say we don't care that we have inconsistent naming in our language.

Hint View is an SSR command, and there used to be a time where SSR was a separate plugin. Should we care because some plugin in the wild might decide to call another vernacular command `Hint MyShinyCommand` that has some random behaviour w.r.t. scoping?

**Comment by RalfJung** [comment]
> It's up to the user to transform the warnings they care about into hard errors. Imposing Local inside sections if the only reason is that the user might have forgotten this was the default is not nice.

In terms of teachability, I think it is pretty nice actually. I am not concerned about power users, I am concerned about novices that have no idea which warning flags to set to learn about these things.

Powerusers know how to change defaults, so defaults have to work for novices. With that in mind, it IMO makes much more sense to error by default, perhaps with some flag to turn the error into a warning or turn it off.

If you truly want to go with warn-by-default, then at least PG's warning needs to be significantly improved. Currently, when I step to somewhere in the middle of the file, it just shows the warnings that accumulate until then in the bottom-right window, with no indication which part of the code the warning is even referring to.

> Hint View is an SSR command, and there used to be a time where SSR was a separate plugin. Should we care because some plugin in the wild might decide to call another vernacular command Hint MyShinyCommand that has some random behaviour w.r.t. scoping?

You should care about the `Hint` vernaculars you are shipping with Coq. As you said yourself, ssreflect "used to be" separate -- the emphasis here is on past tense.

**Comment by ppedrot** [comment]
> You should care about the Hint vernaculars you are shipping with Coq.

Then in light of this point of view, I advocate for renaming `Hint View`.

**Comment by Zimmi48** [comment]
>PG's warning needs to be significantly improved

We want users to take notice of all the warnings ideally, so yes, if this is your experience, please get in touch with PG maintainers (e.g. open an issue) to raise this problem.

**Comment by gares** [comment]
> turn standard `Hint` commands local by default;

Before considering this, I'd like to see stats.

Because I'm afraid that in spite of being safer (for some definition of safe) it would not be the most frequent setting. So you would need to teach the beginner that, now that he has two files, he needs to change all his hints, and I'm pretty sure he would then argue the default is wrong.

**Comment by gares** [comment]
About the warnings, maybe PG is bad, but just look at this @RalfJung : https://github.com/LPCIC/coq-elpi/actions/runs/515206394 , GH is giving me pointers to all the files having warnings in CI! (and I did nothing special).

**Comment by SkySkimmer** [comment]
Can we expect a PR soon?

**Comment by silene** [comment]
I don't remember. What was the rationale for going through an error stage? Why not directly go to the new semantics? I may be overly optimistic, but I feel that most developments that use unadorned `Hint` would work whichever the semantics are. So, breaking them during one release seems a bit gratuitous.

**Comment by ppedrot** [comment]
@silene it's to force the user to realize there is a problem and get a version that is fully incompatible with both behaviours. This can serve as a reference for transition then. (Also, I see a lot of devs in the CI that vomit this warning thousands of times, not sure we're any close to doing this.)

**Comment by silene** [comment]
If the user library breaks after the change of semantics, the users will notice that there is a problem anyway. But if the user library would not break, then we just gratuitously broke the library by making it "fully incompatible". Full incompatibility should never be an end goal in itself. We need a better motivation to force such a breakage upon users.

**Comment by ppedrot** [comment]
> But if the user library would not break, then we just gratuitously broke the library by making it "fully incompatible".

There is literally no example of a library that would behave the same in both directions. It's way too violent a change not to be explicit about it.

**Comment by Zimmi48** [comment]
And note that one way of making the library build again is simply to disable the warning (even if it is fatal by default, it should still be possible to disable it).

**Comment by silene** [comment]
> There is literally no example of a library that would behave the same in both directions.

Could you explain why? I was under the impression that the libraries that will break are libraries that do `Require` but not `Import`. This seems like it should hardly ever happen. (Or do people not import libraries nowadays?)

**Comment by Zimmi48** [comment]
The problem is that `Require` is transitive while `Import` isn't. Therefore, you often end up depending on hints that are far in the transitive-loading chain without realizing.

**Comment by Zimmi48** [comment]
This was the whole point of changing the default locality to `export` BTW.

**Comment by Zimmi48** [comment]
If we just bump the milestone and do not have anyone scheduling to do the work shortly after the 8.15 branching point, then it risks to just not happen until the next version is in sight again.

**Comment by ppedrot** [comment]
IIUC this is still not going to be solved in 8.16 then...

**Comment by ppedrot** [comment]
Postponing.

**Comment by Zimmi48** [comment]
For something that has been postponed three times, it looks like our use of the blocker label is a bit abusive...
